pinot-common/src/main/java/org/apache/pinot/common/config/instance/InstanceAssignmentConfigUtils.java

```diff
@@ -33,58 +33,74 @@
   private InstanceAssignmentConfigUtils() {
   }
 
+  /**
+   * Returns whether the instance assignment is allowed for the given table config.
+   */
+  public static boolean allowInstanceAssignment(TableConfig tableConfig,
+      InstancePartitionsType instancePartitionsType) {
+    TableType tableType = tableConfig.getTableType();
+    Map<InstancePartitionsType, InstanceAssignmentConfig> instanceAssignmentConfigMap =
+        tableConfig.getInstanceAssignmentConfigMap();
+    switch (instancePartitionsType) {
+      // Allow OFFLINE instance assignment if the OFFLINE table has it configured or (for backward-compatibility) is
+      // using replica-group segment assignment
+      case OFFLINE:
+        return tableType == TableType.OFFLINE && ((instanceAssignmentConfigMap != null && instanceAssignmentConfigMap
+            .containsKey(InstancePartitionsType.OFFLINE))
+            || AssignmentStrategy.REPLICA_GROUP_SEGMENT_ASSIGNMENT_STRATEGY
+            .equalsIgnoreCase(tableConfig.getValidationConfig().getSegmentAssignmentStrategy()));
+      // Allow CONSUMING/COMPLETED instance assignment if the REALTIME table has it configured
+      case CONSUMING:
+      case COMPLETED:
+        return tableType == TableType.REALTIME && (instanceAssignmentConfigMap != null && instanceAssignmentConfigMap
+            .containsKey(instancePartitionsType));
+      default:
+        throw new IllegalStateException();
+    }
+  }
+
+  /**
+   * Extracts or generates default instance assignment config from the given table config.
+   */
   public static InstanceAssignmentConfig getInstanceAssignmentConfig(TableConfig tableConfig,
       InstancePartitionsType instancePartitionsType) {
+    Preconditions.checkState(allowInstanceAssignment(tableConfig, instancePartitionsType),
+        "Instance assignment is not allowed for the given table config");
```

Comment:
Can you add `tableName` and `instancePartitionType` to the log?

Reply:
The caller (who handles the exception) will log these information.